r/R/dplyr.R

```diff
@@ -24,6 +24,21 @@ arrow_dplyr_query <- function(.data) {
   # RecordBatch, or Dataset) and the state of the user's dplyr query--things
   # like selected columns, filters, and group vars.
   # An arrow_dplyr_query can contain another arrow_dplyr_query in .data
+
+  supported <- c(
+    "Dataset", "RecordBatch", "RecordBatchReader",
+    "Table", "arrow_dplyr_query", "data.frame"
+  )
+  if (!inherits(.data, supported)) {
+    stop(
+      "'dataset' must be a ",
```

Comment:
The parameter name here is `.data` rather than `'dataset'`.  Can you rework this so that the error message is correct whether it's triggered by `write_dataset()` or some other bit of code calling `arrow_dplyr_query()`?  The first thing that comes to mind is catching the error again in `write_dataset()` though that feels a bit inelegant and I don't think it's the best solution.

Reply:
`substitute(.data, parent.frame())` is probably good enough since, as you observe, `arrow_dplyr_query()` is only ever called from another function, so we'd want to get its argument name from where it was called. 

An alternative, if you weren't comfortable with that, would be to not name the input in the error message. Say something like "A query can only be built with a Dataset, RecordBatch, ...".

The validation should be in `arrow_dplyr_query()` not `write_dataset()` because the input type restriction is about what we can do with a query, not just writing.

Reply:
Unfortunately I don't think `substitute(.data, parent.frame())` (or a variation thereof) is going to be very reliable, as it gives the name of the object passed to the argument in the calling function, not the formal name of the argument. I think it may be easiest to just not name the input. Maybe something like "You must supply a Dataset, RecordBatch..., not a ..."?